src/wcl/doc.h

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2022 SiFive, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You should have received a copy of LICENSE.Apache2 along with
+ * this software. If not, you may obtain a copy at
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cassert>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+namespace wcl {
+
+// Internal state for a doc object. It is intentionally not exposed in the
+// public API and should not be used directly.
+struct doc_state {
+  size_t newline_count = 0;
+  size_t first_width = 0;
+  size_t last_width = 0;
+
+  // Returns the merged state of this and other.
+  doc_state merged(doc_state other) {
+    doc_state merged;
+
+    merged.newline_count = newline_count + other.newline_count;
+
+    merged.last_width = other.last_width;
+    // If other doesn't have any newlines, then the left side needs to be accounted for
+    if (other.newline_count == 0) {
+      merged.last_width += last_width;
+    }
+
+    merged.first_width = first_width;
+    // If the left side doesn't have any newlines, then other side needs to be accounted for
+    if (newline_count == 0) {
+      merged.first_width += other.first_width;
+    }
+
+    return merged;
+  }
+
+  static doc_state from_string(const std::string& str) {
+    doc_state state;
+    for (auto& c : str) {
+      if (c == '\n') {
+        state.last_width = 0;
+        state.newline_count++;
+        continue;
+      }
+
+      if (state.newline_count == 0) {
+        state.first_width++;
+      }
+
+      state.last_width++;
+    }
+    return state;
+  }
+};
+
+class doc_impl_base {
+ protected:
+  // On construction this has to be set
+  size_t size_;
+
+  doc_state state_;
+
+  doc_impl_base(size_t size) : size_(size) {}
+
+ public:
+  virtual ~doc_impl_base() = default;
+
+  // methods that all RopeImpl share
+  virtual void write(std::ostream&) const = 0;
+  size_t size() const { return size_; }
+  size_t first_width() const { return state_.first_width; }
+  size_t last_width() const { return state_.last_width; }
+  size_t newline_count() const { return state_.newline_count; }
+  bool has_newline() const { return state_.newline_count > 0; }
+  doc_state state() const { return state_; }
```

Comment:
you can leave this out and state_ can just be accessed directly.

Reply:
I don't actually think this works because I need to do something like `left->state_` in the pair impl. Even though I'm inside of a subclass, I think I can only see _my_ protected members

Reply:
nope, you can see any protected member from the same class.

Reply:
Well its not compiling :)

```
In file included from tools/wake-format/wake-format.cpp:23:
src/wcl/doc.h: In constructor 'wcl::doc_impl_pair::doc_impl_pair(std::shared_ptr<wcl::doc_impl_base>, std::shared_ptr<wcl::doc_impl_base>)':
src/wcl/doc.h:128:29: error: 'wcl::doc_state wcl::doc_impl_base::state_' is protected within this context
       : doc_impl_base(left->state_.merged(right->state_)),
                             ^~~~~~
src/wcl/doc.h:93:13: note: declared protected here
   doc_state state_;
             ^~~~~~
src/wcl/doc.h:128:50: error: 'wcl::doc_state wcl::doc_impl_base::state_' is protected within this context
       : doc_impl_base(left->state_.merged(right->state_)),
                                                  ^~~~~~
src/wcl/doc.h:93:13: note: declared protected here
   doc_state state_;
```

also doc.h needs to forward `state` to rope_builder, so even if it worked above it wouldn't work for doc which isn't a subclass

Reply:
I'll just go ahead and merge this. Will revisit later if there is a way to do it without the function.

Reply:
oh huh...today I was one of the lucky 10k...and was wrong lol